src/hardware/opl.cpp

```diff
@@ -473,44 +203,104 @@ uint8_t Chip::Read()
 	return ret;
 }
 
-void OPL::Init(const uint16_t sample_rate)
+uint8_t OplChip::EsfmReadbackReg(const uint16_t reg)
+{
+	uint8_t ret = 0;
+
+	switch (reg) {
+	case 0x02: ret = timer0.GetCounter(); break;
+	case 0x03: ret = timer1.GetCounter(); break;
+	case 0x04:
+		ret = timer0.IsEnabled() & 1;
+		ret |= (timer1.IsEnabled() & 1) << 1;
+		ret |= (timer1.IsMasked() & 1) << 5;
+		ret |= (timer1.IsMasked() & 1) << 6;
+	}
+	return ret;
+}
+
+void OPL::Init()
 {
-	newm = 0;
-	OPL3_Reset(&oplchip, sample_rate);
+	opl.newm = 0;
+
+	if (opl_mode == OplMode::Esfm) {
+		ESFM_init(&esfm.chip);
+	} else {
+		OPL3_Reset(&opl.chip, OplSampleRateHz);
+	}
 
-	ms_per_frame = millis_in_second / sample_rate;
+	ms_per_frame = millis_in_second / OplSampleRateHz;
 
 	memset(cache, 0, ARRAY_LEN(cache));
 
-	switch (mode) {
-	case Mode::Opl3: break;
-	case Mode::Opl3Gold:
-		adlib_gold = std::make_unique<AdlibGold>(sample_rate);
-		break;
-	case Mode::Opl2: break;
-	case Mode::DualOpl2:
+	switch (opl_mode) {
+	case OplMode::Opl2: break;
+
+	case OplMode::DualOpl2:
 		// Setup opl3 mode in the hander
 		WriteReg(0x105, 1);
+
 		// Also set it up in the cache so the capturing will start opl3
 		CacheWrite(0x105, 1);
 		break;
+
+	case OplMode::Opl3: break;
+
+	case OplMode::Opl3Gold:
+		adlib_gold = std::make_unique<AdlibGold>(OplSampleRateHz);
+		break;
+
+	case OplMode::Esfm: break;
+
+	default:
+		assertm(false,
+		        format_str("Invalid OPL mode: %s", to_string(opl_mode)));
 	}
 }
 
 void OPL::WriteReg(const io_port_t selected_reg, const uint8_t val)
 {
-	OPL3_WriteRegBuffered(&oplchip, selected_reg, val);
-	if (selected_reg == 0x105)
-		newm = selected_reg & 0x01;
+	if (opl_mode == OplMode::Esfm) {
+		ESFM_write_reg_buffered_fast(&esfm.chip, selected_reg, val);
+
+
+	} else { // OPL
+		OPL3_WriteRegBuffered(&opl.chip, selected_reg, val);
+		if (selected_reg == 0x105) {
+			opl.newm = selected_reg & 0x01;
+		}
+	}
 }
 
+// TODO Add OPL::ReadbackReg for ESFM
+
 io_port_t OPL::WriteAddr(const io_port_t port, const uint8_t val)
 {
-	io_port_t addr = val;
-	if ((port & 2) && (addr == 0x05 || newm)) {
-		addr |= 0x100;
+	if (opl_mode == OplMode::Esfm) {
```

Comment:
I considered it, but no. They are too intertwined, and a separate class would solve nothing, it would be just a different kind of complexity. I'd say it would be worse.